sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala

```diff
@@ -230,12 +230,28 @@ case class InsertIntoHiveTable(
         var doHiveOverwrite = overwrite
 
         if (oldPart.isEmpty || !ifPartitionNotExists) {
+          // SPARK-29295: When insert overwrite to a Hive external table partition, if the
+          // partition does not exist, Hive will not check if the external partition directory
+          // exists or not before copying files. So if users drop the partition, and then do
+          // insert overwrite to the same partition, the partition will have both old and new
+          // data.
+          val updatedPart = if (overwrite && table.tableType == CatalogTableType.EXTERNAL) {
+            AlterTableAddPartitionCommand(
```

Comment:
Instead of constructing partition directory manually, I let hive add the target parititon to overwrite. We will delete the partition directory below.

Reply:
How can we rollback if the follow `loadPartition` operation fails?

Reply:
Do we provide rollback for load partition before?